### PR TITLE
Add multiple acquisition methods for AL

### PR DESCRIPTION
**Pull request by y0ast**
Previously only entropy scoring was implemented, this PR adds uniform, margin and density based AL methods. We refactored a few of the AL functions to make the multi acquisition code simpler.

The code introduces early stopping, but it's unclear if we want to continue with this. The general big next step is to figure out training on a small dataset. This PR introduces some initial prints, which will be cleaned up when we figured out a nice setup.

cc @zi-w @dusenberrymw 

All co-authored-by: @BlackHC

### CODE REVIEW

File: baselines/jft/active_learning.py [hunk 1]
```diff
@@ -206,6 +318,10 @@ def finetune(*,
              ds,
              rngs_loop,
              total_steps,
+             train_ds,
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:321]: Add `train_ds` to the docstring below. Also, I think it would helpful to name this `train_eval_ds` to distinguish from a normal shuffled training dataset.

File: baselines/jft/active_learning.py [hunk 2]
```diff
@@ -489,93 +633,167 @@ def main(config, output_dir):
...
+        process_batch_size=local_batch_size,
+        preprocess_fn=preprocess_spec.parse(
+            spec=config.pp_train, available_ops=preprocess_utils.all_ops()),
+        shuffle_buffer_size=config.shuffle_buffer_size,
+        prefetch_size=config.get("prefetch_to_host", 2),
+        # TODO: double check if below actually leads to bootstrap sampling
+        num_epochs=num_repeats,
+    )
+
+    train_ds = input_utils.get_data(
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:697]: I think it would be helpful to rename this to something like `train_eval_ds` since it's intended to be used for eval only (e.g., isn't shuffled and is using the eval preprocessing (`pp_eval`)).

File: baselines/jft/active_learning_test.py [hunk 3]
```diff
@@ -69,15 +69,24 @@ def test_active_learning_script(self):
     config.train_split = 'train[:98%]'
     config.batch_size = 8
     config.max_labels = 3
-    config.acquisition_batch_size = 2
+    config.acquisition_batch_size = 5
+    config.total_steps = 6
     config.dataset_dir = data_dir
     config.model_init = checkpoint_path
+    acquisition_method = 'uniform'
```
Comment (dusenberrymw) [inline baselines/jft/active_learning_test.py:76]: Can you test all of the new acquisition functions? Best option would be to use same kind of `@parameterized.parameters` setup that [deterministic_test.py](https://github.com/google/uncertainty-baselines/blob/main/baselines/jft/deterministic_test.py#L44) uses, with two parameters for this `test_active_learning_script` function: `acquisition_method` and `correct_ids`. One line for each of the four new acquisition methods.

File: baselines/jft/active_learning.py [hunk 4]
```diff
@@ -52,109 +52,222 @@
...
+import ood_utils  # local file import from baselines.jft
 import preprocess_utils  # local file import from baselines.jft
 import train_utils  # local file import from baselines.jft
 
 config_flags.DEFINE_config_file(
     "config", None, "Training configuration.", lock_config=True)
 flags.DEFINE_string("output_dir", default=None, help="Work unit directory.")
+flags.DEFINE_enum("acquisition_method", default="uniform",
+                  enum_values=["uniform", "density", "margin", "entropy"],
+                  help="Work unit directory.")
```
Comment (dusenberrymw) [inline baselines/jft/active_learning.py:64]: Update the help string.

**Review by dusenberrymw — CHANGES_REQUESTED** [review]
Thanks, Joost! Left comments, and can you also squash the commits and add `Co-authored-by: Andreas Kirsch <blackhc@gmail.com>` to the end of the commit description?

File: baselines/jft/active_learning_test.py (hunk 3 above)
Comment (y0ast) [reply baselines/jft/active_learning_test.py:76]: I tried to do this, but I couldn't figure out how to do the test.

The first acquisition is always done with `uniform` so to test the others we would need to first acquire with `uniform`, then train, then acquire with the other methods. Unfortunately the train step always crashes since we filter by id, but the ids are randomized (by `tfds.testing.mock_data`) so we end up with an empty training set.

**Comment by y0ast** [comment]
Thanks for the review Mike!

I'll squash and add the co-authored-by commit line after we figured out the test problem (so we can track how/if comments are addressed).

**Comment by y0ast** [comment]
I fixed the tests and learned about: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision

JAX was silently turning my python `int` into `int32` (by just dropping bits it seems), and if you generate random bytes as id and interpret as int actually leads to id > MAX_INT32 surprisingly often.

**Review by dusenberrymw — APPROVED** [review]
LGTM. Thanks, Joost (& Andreas)!
